modules/fedimint-meta-server/src/lib.rs

```diff
@@ -0,0 +1,509 @@
+        peers
+            .iter()
+            .map(|&peer| {
+                let config = MetaConfig {
+                    local: MetaConfigLocal {},
+                    private: MetaConfigPrivate,
+                    consensus: MetaConfigConsensus {},
+                };
+                (peer, config.to_erased())
+            })
+            .collect()
+    }
+
+    /// Generates configs for all peers in an untrusted manner
+    async fn distributed_gen(
+        &self,
+        _peers: &PeerHandle,
+        params: &ConfigGenModuleParams,
+    ) -> DkgResult<ServerModuleConfig> {
+        let _params = self.parse_params(params).unwrap();
+
+        Ok(MetaConfig {
+            local: MetaConfigLocal {},
+            private: MetaConfigPrivate,
+            consensus: MetaConfigConsensus {},
+        }
+        .to_erased())
+    }
+
+    /// Converts the consensus config into the client config
+    fn get_client_config(
+        &self,
+        config: &ServerModuleConsensusConfig,
+    ) -> anyhow::Result<MetaClientConfig> {
+        let _config = MetaConfigConsensus::from_erased(config)?;
+        Ok(MetaClientConfig {})
+    }
+
+    fn validate_config(
+        &self,
+        _identity: &PeerId,
+        _config: ServerModuleConfig,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
+
+    /// DB migrations to move from old to newer versions
+    fn get_database_migrations(&self) -> BTreeMap<DatabaseVersion, ServerMigrationFn> {
+        BTreeMap::new()
+    }
+}
+
+/// Meta module
+#[derive(Debug)]
+pub struct Meta {
+    pub cfg: MetaConfig,
+    pub our_peer_id: PeerId,
+    pub num_peers: NumPeers,
+}
+
+impl Meta {
+    /// Check the difference between what's desired to be submitted and what's
+    /// already submitted or consensus value.
+    ///
+    /// Returns:
+    /// `(items_to_submit, desired_keys_to_delete)`
+    async fn desired_submitted_diff(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_, NonCommittable>,
+    ) -> (Vec<MetaConsensusItem>, Vec<MetaKey>) {
+        let desired: Vec<_> = Self::get_desired(dbtx).await;
+
+        let mut to_delete: Vec<MetaKey> = vec![];
+        let mut to_submit = vec![];
+
+        for (key, MetaDesiredValue { value, salt }) in desired {
+            if Self::get_submission(dbtx, key, self.our_peer_id)
+                .await
+                .as_ref()
+                == Some(&MetaSubmissionValue {
+                    value: value.clone(),
+                    salt,
+                })
+            {
+                to_delete.push(key);
+            } else if Self::get_consensus(dbtx, key).await.as_ref() == Some(&value) {
+                // When a value is new and equal to current consensus, it should both: be
+                // submitted so it can clear/cancel previous submissions, and
+                // deleted (afterwards).
+                to_delete.push(key);
+                to_submit.push(MetaConsensusItem { key, value, salt });
```

Comment:
Unclear why we *need* to submit again, but ok I guess? Others will likely filter it out if it's a dup?

Reply:
Yeah. If it's a dup, it will be ignored. If it isn't it will clear guardian's previous submission to match the consensus.